File: client/app/technical-indicators/macd.service.ts
import { StockPrice } from '../charts/stock-price'
import { MACD } from './macd'

// // http://investexcel.net/how-to-calculate-macd-in-excel/
export class MACDService {
    PARAM_1: number = 12;
    PARAM_2: number = 26;
    PARAM_3: number = 9;
    calculate(stockPrices: StockPrice[]) : MACD[] {
        //debugger;
        let entries = stockPrices.map(sp => new MACD(sp.date, sp.price));
        entries.reduce((previousValue, currentValue, currentIndex) => {
            let index = currentIndex + 1;
            currentValue.priceSum = previousValue.priceSum + currentValue.price;
            currentValue.priceAverage = currentValue.priceSum / index;
            
            // 12 Day EMA
            if (index === this.PARAM_1) {
                currentValue.ema12day = currentValue.priceAverage;
            } else if (index > this.PARAM_1) {
                currentValue.ema12day = currentValue.price * (2/(this.PARAM_1 + 1)) + previousValue.ema12day * (1 - 2/(this.PARAM_1 + 1));
            }

            // 26 Day EMA
            if (index === this.PARAM_2) {
                currentValue.ema26day = currentValue.priceAverage;
            } else if (index > this.PARAM_2) {
                currentValue.ema26day = currentValue.price * (2/(this.PARAM_2 + 1)) + previousValue.ema26day * (1 - 2/(this.PARAM_2 + 1));
            }

            // MCDA
            if (index >= this.PARAM_2) {
                currentValue.macd = currentValue.ema12day - currentValue.ema26day;
                currentValue.macdSum = previousValue.macdSum + currentValue.macd;
                currentValue.macdAverage = currentValue.macdSum / (index - this.PARAM_2 + 1);
            }

            // Signal
            if (index === (this.PARAM_2 + this.PARAM_3 - 1)) {
                currentValue.signal = currentValue.macdAverage;
            } else if (index > (this.PARAM_2 + this.PARAM_3 - 1)) {
                currentValue.signal = currentValue.macd * (2/(this.PARAM_3 + 1)) + previousValue.signal * (1 - 2/(this.PARAM_3 + 1));
            }

            // Histogram
            if (index >= (this.PARAM_2 + this.PARAM_3 - 1)) {
                currentValue.histogram = currentValue.macd - currentValue.signal;
            }

            return currentValue;
        }, new MACD(new Date(1900, 1, 1), 0));

        return entries;
    }
}

File: server/RuleOne/src/InvSys.RuleOne.Core/Services/ThreeTools/MACDService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using InvSys.RuleOne.Core.Models.ThreeTools;
using InvSys.StockQuotes.Api.Client.Proxy.Models;

namespace InvSys.RuleOne.Core.Services.ThreeTools
{
    // MACD - Moving Average Convergence Divergence
    // http://investexcel.net/how-to-calculate-macd-in-excel/
    public class MACDService : IMACDService
    {
        public ICollection<MACDData> Calculate(ICollection<HistoricalQuote> prices, int param1, int param2, int param3)
        {
            var entries = prices.Select(p => new MACDData { Date = p.Date.Value, Price = (decimal)p.Close.Value }).ToList(); // TODO: swagger and decimal (https://github.com/Azure/autorest/issues/83)
            entries.Insert(0, new MACDData { Date = new DateTime(1900, 1, 1), Price = 0 });
            for (var index = 1; index < entries.Count; index++)
            {
                var previousValue = entries[index - 1];
                var currentValue = entries[index];

                currentValue.PriceSum = previousValue.PriceSum + currentValue.Price;
                currentValue.PriceAverage = currentValue.PriceSum / index;

                // 12 Day EMA
                if (index == param1)
                {
                    currentValue.EMA12Day = currentValue.PriceAverage;
                }
                else if (index > param1)
                {
                    currentValue.EMA12Day = currentValue.Price * ((decimal)2 / (param1 + 1)) + previousValue.EMA12Day * (1 - (decimal)2 / (param1 + 1));
                }

                // 26 Day EMA
                if (index == param2)
                {
                    currentValue.EMA26Day = currentValue.PriceAverage;
                }
                else if (index > param2)
                {
                    currentValue.EMA26Day = currentValue.Price * ((decimal)2 / (param2 + 1)) + previousValue.EMA26Day * (1 - (decimal)2 / (param2 + 1));
                }

                // MCDA
                if (index >= param2)
                {
                    currentValue.MACD = currentValue.EMA12Day - currentValue.EMA26Day;
                    currentValue.MACDSum = previousValue.MACDSum + currentValue.MACD;
                    currentValue.MACDAverage = currentValue.MACDSum / (index - param2 + 1);
                }

                // Signal
                if (index == (param2 + param3 - 1))
                {
                    currentValue.Signal = currentValue.MACDAverage;
                }
                else if (index > (param2 + param3 - 1))
                {
                    currentValue.Signal = currentValue.MACD * ((decimal)2 / (param3 + 1)) + previousValue.Signal * (1 - (decimal)2 / (param3 + 1));
                }

                // Histogram
                if (index >= (param2 + param3 - 1))
                {
                    currentValue.Histogram = currentValue.MACD- currentValue.Signal;
                }
            }

            return entries.Where(e => e.MACD != 0 && e.Signal != 0).OrderBy(e => e.Date).ToList();
        }
    }
}
